### PR TITLE
Add vnc server app

### PR DESCRIPTION
**Pull request by AlanGriffiths**
To use this:

```
ssh <username>@<hostname> -- egmde-confined-desktop.vnc&
ssh -N -T -L 5900:localhost:5900 <username>@<hostname>&
gvncviewer localhost:0
```

### CODE REVIEW

**Comment by AlanGriffiths** [comment]
Note: If using a qemu VM set up according to [this tutorial](https://ubuntu.com/tutorials/ubuntu-core-preparing-a-virtual-machine-with-graphics-support) then replace `<username>@<hostname>` with `-p 10022 <username>@localhost`

Note1: If installing on a "classic" system it is necessary to _disconnect_ the x11 interface.

File: glue/bin/run-vnc [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+xvfb_pid=$!
+
+trap "kill $xvfb_pid" INT TERM
+
+until [ -O "${XDG_RUNTIME_DIR}/${WAYLAND_DISPLAY}" ]
+do
+  inotifywait -qq --timeout 5 --event create $(dirname "${XDG_RUNTIME_DIR}/${WAYLAND_DISPLAY}")
+done
+
+env -u WAYLAND_DISPLAY x11vnc -display ${DISPLAY} -bg -nopw -listen localhost -xkb -auth ${x11_auth_file} -ncache 10
```
Comment (Saviq) [inline glue/bin/run-vnc]: Why not listen on 0? You could require a password to be set for security, but it would reduce the need for the ssh tunnel?

**Comment by Saviq** [comment]
Also:

```
$ ssh michal-laptop.local -- egmde-confined-desktop.vnc
+ mkdir -p /run/user/1000/snap.egmde-confined-desktop -m 700
+ mkdir -p /home/michal/snap/egmde-confined-desktop/common/.cache
+ DISPLAY_NO=66
+ DISPLAY=:66
++ mktemp --tmpdir=/run/user/1000/snap.egmde-confined-desktop
+ x11_auth_file=/run/user/1000/snap.egmde-confined-desktop/tmp.M5Rt05Yxmj
+ '[' -O /run/user/1000/snap.egmde-confined-desktop/ ']'
+ port=0
+ '[' -e /run/user/1000/snap.egmde-confined-desktop/wayland-0 ']'
+ WAYLAND_DISPLAY=wayland-0
+ xvfb_pid=40511
+ trap 'kill 40511' INT TERM
+ '[' -O /run/user/1000/snap.egmde-confined-desktop/wayland-0 ']'
+ xvfb-run --auth-file=/run/user/1000/snap.egmde-confined-desktop/tmp.M5Rt05Yxmj --server-num=66 /snap/egmde-confined-desktop/6460/bin/run-egmde
++ dirname /run/user/1000/snap.egmde-confined-desktop/wayland-0
+ inotifywait -qq --timeout 5 --event create /run/user/1000/snap.egmde-confined-desktop
xvfb-run: error: Xvfb failed to start
+ '[' -O /run/user/1000/snap.egmde-confined-desktop/wayland-0 ']'
++ dirname /run/user/1000/snap.egmde-confined-desktop/wayland-0
+ inotifywait -qq --timeout 5 --event create /run/user/1000/snap.egmde-confined-desktop
```

**Comment by AlanGriffiths** [comment]
Huh?
```
xvfb-run: error: Xvfb failed to start
```

I wonder what is different on your system?

File: glue/bin/run-vnc (hunk 1 above)
Comment (AlanGriffiths) [reply glue/bin/run-vnc]: Habit mostly, it is "how it ought to be done" in this hostile world.

Comment (Saviq) [reply glue/bin/run-vnc]: OK.

**Comment by Saviq** [comment]
> I wonder what is different on your system?

The `x11` interface connection, as you found. Maybe fail early if `x11` is connected?

"X11VNC Server" is an app in the app launcher, should maybe drop its `.desktop` file?

**Comment by AlanGriffiths** [comment]
> "X11VNC Server" is an app in the app launcher, should maybe drop its `.desktop` file?

Yes, coming up

File: glue/bin/run-vnc [hunk 2]
```diff
@@ -0,0 +1,42 @@
+#!/bin/bash
+set -ex
+
+if snapctl is-connected x11; then
+  echo Cannot initialize VNC environment please disconnect $SNAP_NAME:x11
```
Comment (Saviq) [inline glue/bin/run-vnc]:
```suggestion
  echo Cannot initialize VNC environment, please disconnect $SNAP_INSTANCE_NAME:x11
```

File: glue/bin/run-vnc [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+xvfb_pid=$!
+
+trap "kill $xvfb_pid" INT TERM
+
+until [ -O "${XDG_RUNTIME_DIR}/${WAYLAND_DISPLAY}" ]
+do
+  inotifywait -qq --timeout 5 --event create $(dirname "${XDG_RUNTIME_DIR}/${WAYLAND_DISPLAY}")
+done
+
+env -u WAYLAND_DISPLAY x11vnc -display ${DISPLAY} -bg -nopw -listen localhost -xkb -auth ${x11_auth_file} -ncache 10 -forever -alwaysshared
```
Comment (Saviq) [inline glue/bin/run-vnc:40]: How about this?
```suggestion
env -u WAYLAND_DISPLAY x11vnc -display ${DISPLAY} -bg -nopw -listen localhost -xkb -auth ${x11_auth_file} -ncache 10 -forever -alwaysshared "$@"
```

**Review by Saviq — APPROVED** [review]
Just a couple suggestions inline.

bors delegate

File: glue/bin/run-vnc (hunk 3 above)
Comment (AlanGriffiths) [reply glue/bin/run-vnc:40]: That's weird. We already pass the egmde invocation to `xvfb-run` (line 30), does it even make sense to pass it to `x11vnc`?

**Comment by AlanGriffiths** [comment]
bors merge

File: glue/bin/run-vnc (hunk 3 above)
Comment (Saviq) [reply glue/bin/run-vnc:40]: Missed it! And yeah it makes more sense to pass there than to x11vnc, even if the command is `vnc`…

**Pull request merged**